internals/daemon/daemon.go

```diff
@@ -194,7 +202,31 @@ func (c *Command) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	if rspe := access.CheckAccess(c.d, r, ucred, user); rspe != nil {
+	// Optimisation: avoid calling userFromRequest, which acquires the state
+	// lock, in case we don't need to (when endpoint is OpenAccess). This
+	// avoids holding the state lock for /v1/health in particular, which is
+	// not good: https://github.com/canonical/pebble/pull/369
+	var user *UserState
+	if _, isOpen := access.(OpenAccess); !isOpen {
+		user, err = userFromRequest(c.d.state, r, ucred)
+		if err != nil {
+			Forbidden("forbidden").ServeHTTP(w, r)
+			return
+		}
+	}
+
+	// If we don't have a named-identity user, use ucred UID to see if we have a default.
+	if user == nil && ucred != nil {
+		if ucred.Uid == 0 || ucred.Uid == uint32(os.Getuid()) {
+			// Admin if UID is 0 (root) or the UID the daemon is running as.
+			user = &UserState{Access: state.AdminAccess, UID: &ucred.Uid}
+		} else {
+			// Regular read access if any other local UID.
+			user = &UserState{Access: state.ReadAccess, UID: &ucred.Uid}
+		}
+	}
```

Comment:
Shouldn't this be in `userFromRequest()`? Seems like this is for backwards compatibility? Or I guess so that we can make the optimization above that `userFromRequest()` is called only conditionally?

Reply:
Yep, I've moved this to `userFromRequest` -- good call.